test/common/config/subscription_factory_test.cc

```diff
@@ -0,0 +1,127 @@
+      "envoy::api::v2::ConfigSource must have a singleton cluster name specified");
+}
+
+TEST_F(SubscriptionFactoryTest, FilesystemSubscription) {
+  envoy::api::v2::ConfigSource config;
+  config.set_path("/blahblah");
+  auto* watcher = new Filesystem::MockWatcher();
+  EXPECT_CALL(dispatcher_, createFilesystemWatcher_()).WillOnce(Return(watcher));
+  EXPECT_CALL(*watcher, addWatch("/blahblah", _, _));
+  subscriptionFromConfigSource(config)->start({"foo"}, callbacks_);
+}
+
+TEST_F(SubscriptionFactoryTest, LegacySubscription) {
+  envoy::api::v2::ConfigSource config;
+  auto* api_config_source = config.mutable_api_config_source();
+  api_config_source->set_api_type(envoy::api::v2::ApiConfigSource::REST_LEGACY);
+  api_config_source->add_cluster_name("eds_cluster");
+  EXPECT_CALL(*legacy_subscription_, start(_, _));
+  subscriptionFromConfigSource(config)->start({"foo"}, callbacks_);
+}
+
+TEST_F(SubscriptionFactoryTest, HttpSubscription) {
+  envoy::api::v2::ConfigSource config;
+  auto* api_config_source = config.mutable_api_config_source();
+  api_config_source->set_api_type(envoy::api::v2::ApiConfigSource::REST);
+  api_config_source->add_cluster_name("eds_cluster");
+  EXPECT_CALL(dispatcher_, createTimer_(_));
+  EXPECT_CALL(cm_, httpAsyncClientForCluster("eds_cluster"));
+  EXPECT_CALL(cm_.async_client_, send_(_, _, _))
+      .WillOnce(Invoke([this](Http::MessagePtr& request, Http::AsyncClient::Callbacks& callbacks,
+                              const Optional<std::chrono::milliseconds>& timeout) {
+        UNREFERENCED_PARAMETER(callbacks);
+        UNREFERENCED_PARAMETER(timeout);
+        EXPECT_EQ("POST", std::string(request->headers().Method()->value().c_str()));
+        EXPECT_EQ("eds_cluster", std::string(request->headers().Host()->value().c_str()));
+        EXPECT_EQ("/v2/discovery:endpoints",
+                  std::string(request->headers().Path()->value().c_str()));
+        return &http_request_;
+      }));
+  EXPECT_CALL(http_request_, cancel());
+  subscriptionFromConfigSource(config)->start({"foo"}, callbacks_);
+}
+
+TEST_F(SubscriptionFactoryTest, GrpcSubscription) {
+  envoy::api::v2::ConfigSource config;
+  auto* api_config_source = config.mutable_api_config_source();
+  api_config_source->set_api_type(envoy::api::v2::ApiConfigSource::GRPC);
+  api_config_source->add_cluster_name("eds_cluster");
+  EXPECT_CALL(dispatcher_, createTimer_(_));
+  EXPECT_CALL(cm_, httpAsyncClientForCluster("eds_cluster"));
+  NiceMock<Http::MockAsyncClientStream> stream;
+  EXPECT_CALL(cm_.async_client_, start(_, _)).WillOnce(Return(&stream));
+  Http::TestHeaderMapImpl headers{
+      {":method", "POST"},
+      {":path", "/envoy.api.v2.EndpointDiscoveryService/StreamEndpoints"},
+      {":authority", "eds_cluster"},
+      {"content-type", "application/grpc"}};
+  EXPECT_CALL(stream, sendHeaders(HeaderMapEqualRef(&headers), _));
+  subscriptionFromConfigSource(config)->start({"foo"}, callbacks_);
+}
+
+} // namespace Config
```

Comment:
random note: I think at this point we should be able to easily upgrade clang-format and use 5.0 due to how we have CI setup. We should just upgrade sometime soon and then we can fix some other inconsistencies.